primitives/consensus/common/src/lib.rs

```diff
@@ -71,14 +74,54 @@ pub enum BlockStatus {
 /// Environment producer for a Consensus instance. Creates proposer instance and communication streams.
 pub trait Environment<B: BlockT> {
 	/// The proposer type this creates.
-	type Proposer: Proposer<B>;
+	type Proposer: Proposer<B> + 'static;
 	/// Error which can occur upon creation.
-	type Error: From<Error>;
+	type Error: From<Error> + std::fmt::Debug + 'static;
 
 	/// Initialize the proposal logic on top of a specific header. Provide
 	/// the authorities at that header.
-	fn init(&mut self, parent_header: &B::Header)
-		-> Result<Self::Proposer, Self::Error>;
+	fn init(&mut self, parent_header: &B::Header) -> Result<Self::Proposer, Self::Error>;
+}
+
+/// A proposal that is created by a [`Proposer`].
+pub struct Proposal<Block: BlockT, Transaction> {
+	/// The block that was build.
+	pub block: Block,
+	/// Optional proof that was recorded while building the block.
+	pub proof: Option<state_machine::StorageProof>,
+	/// The storage changes while building this block.
+	pub storage_changes: state_machine::StorageChanges<Transaction, HasherFor<Block>, NumberFor<Block>>,
+}
+
+/// Used as parameter to [`Proposer`] to tell the requirement on recording a proof.
+///
+/// When `RecordProof::Yes` is given, all accessed trie nodes should be saved. These recorded
+/// trie nodes can be used by a third party to proof this proposal without having access to the
+/// full storage.
+#[derive(Copy, Clone, PartialEq)]
+pub enum RecordProof {
+	Yes,
```

Comment:
Doesn't this trigger a warning about missing docs?

Reply:
No.

Reply:
But there is also no warn on no docs attribute for this file.

Reply:
Hah, right. I guess that's explains it :)